### ISSUE TITLE
`final` keyword in `instanceof` comparisons causes parse error

### ISSUE THREAD

**Issue by SimplyDanny** [checkstyle/checkstyle#13071]
## Java Compiler (fails, see description below)
```bash
/var/tmp $ javac Test.java
Test.java:6: error: modifier final not allowed here
                if (n instanceof final Integer) {}
                                       ^
1 error
```

## Configuration (doesn't matter)
```bash
/var/tmp $ cat config.xml
```
```xml
<!DOCTYPE module PUBLIC
  "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
  "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="AvoidStarImport">
            <property name="severity" value="warning" />
        </module>
    </module>
</module>
```
## Java Source
```bash
/var/tmp $ cat Test.java
```
```java
package org.checkstyle.test;

class Test {
    void check() {
        Number n = 1;
        if (n instanceof final Integer) {}
    }
}
```

## Checkstyle Error
```bash
/var/tmp $ java $RUN_LOCALE -jar checkstyle-10.11.0-all.jar -c config.xml YOUR_FILE.java
```
```
Starting audit...
com.puppycrawl.tools.checkstyle.api.CheckstyleException: Exception was thrown while processing Test.java
        at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:306)
        at com.puppycrawl.tools.checkstyle.Checker.process(Checker.java:223)
        at com.puppycrawl.tools.checkstyle.Main.runCheckstyle(Main.java:415)
        at com.puppycrawl.tools.checkstyle.Main.runCli(Main.java:338)
        at com.puppycrawl.tools.checkstyle.Main.execute(Main.java:195)
        at com.puppycrawl.tools.checkstyle.Main.main(Main.java:130)
Caused by: com.puppycrawl.tools.checkstyle.api.CheckstyleException: IllegalStateException occurred while parsing file C:\Users\msd1lo\Stuff\checkstyle\Test.java.
        at com.puppycrawl.tools.checkstyle.JavaParser.parse(JavaParser.java:105)
        at com.puppycrawl.tools.checkstyle.TreeWalker.processFiltered(TreeWalker.java:152)
        at com.puppycrawl.tools.checkstyle.api.AbstractFileSetCheck.process(AbstractFileSetCheck.java:98)
        at com.puppycrawl.tools.checkstyle.Checker.processFile(Checker.java:334)
        at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:293)
        ... 5 more
Caused by: java.lang.IllegalStateException: 6:32: mismatched input ')' expecting {IDENT, 'record', 'yield', 'non-sealed', 'sealed', 'permits', 'when'}
        at com.puppycrawl.tools.checkstyle.JavaParser$CheckstyleErrorListener.syntaxError(JavaParser.java:255)
        at org.antlr.v4.runtime.ProxyErrorListener.syntaxError(ProxyErrorListener.java:41)
        at org.antlr.v4.runtime.Parser.notifyErrorListeners(Parser.java:544)
        at org.antlr.v4.runtime.DefaultErrorStrategy.reportInputMismatch(DefaultErrorStrategy.java:327)
        at org.antlr.v4.runtime.DefaultErrorStrategy.reportError(DefaultErrorStrategy.java:139)
        at com.puppycrawl.tools.checkstyle.CheckstyleParserErrorStrategy.recoverInline(CheckstyleParserErrorStrategy.java:38)
        at com.puppycrawl.tools.checkstyle.grammar.java.JavaLanguageParser.id(JavaLanguageParser.java:12336)
        at com.puppycrawl.tools.checkstyle.grammar.java.JavaLanguageParser.typePattern(JavaLanguageParser.java:12228)
        at com.puppycrawl.tools.checkstyle.grammar.java.JavaLanguageParser.primaryPattern(JavaLanguageParser.java:12129)
        at com.puppycrawl.tools.checkstyle.grammar.java.JavaLanguageParser.expr(JavaLanguageParser.java:9914)
        at com.puppycrawl.tools.checkstyle.grammar.java.JavaLanguageParser.expression(JavaLanguageParser.java:8787)
        at com.puppycrawl.tools.checkstyle.grammar.java.JavaLanguageParser.parExpression(JavaLanguageParser.java:8686)
        at com.puppycrawl.tools.checkstyle.grammar.java.JavaLanguageParser.statement(JavaLanguageParser.java:6766)
        at com.puppycrawl.tools.checkstyle.grammar.java.JavaLanguageParser.blockStatement(JavaLanguageParser.java:6195)
        at com.puppycrawl.tools.checkstyle.grammar.java.JavaLanguageParser.block(JavaLanguageParser.java:6098)
        at com.puppycrawl.tools.checkstyle.grammar.java.JavaLanguageParser.methodBody(JavaLanguageParser.java:2937)
        at com.puppycrawl.tools.checkstyle.grammar.java.JavaLanguageParser.methodDeclaration(JavaLanguageParser.java:2895)
        at com.puppycrawl.tools.checkstyle.grammar.java.JavaLanguageParser.memberDeclaration(JavaLanguageParser.java:2742)
        at com.puppycrawl.tools.checkstyle.grammar.java.JavaLanguageParser.classBodyDeclaration(JavaLanguageParser.java:2668)
        at com.puppycrawl.tools.checkstyle.grammar.java.JavaLanguageParser.classBody(JavaLanguageParser.java:2474)
        at com.puppycrawl.tools.checkstyle.grammar.java.JavaLanguageParser.classDeclaration(JavaLanguageParser.java:1093)
        at com.puppycrawl.tools.checkstyle.grammar.java.JavaLanguageParser.types(JavaLanguageParser.java:750)
        at com.puppycrawl.tools.checkstyle.grammar.java.JavaLanguageParser.typeDeclaration(JavaLanguageParser.java:664)
        at com.puppycrawl.tools.checkstyle.grammar.java.JavaLanguageParser.compilationUnit(JavaLanguageParser.java:411)
        at com.puppycrawl.tools.checkstyle.JavaParser.parse(JavaParser.java:99)
        ... 9 more
Caused by: org.antlr.v4.runtime.InputMismatchException
        ... 29 more
Checkstyle ends with 1 errors.
```

---

## Description

`javac` isn't able to compile `Test.java` reporting that the `final` keyword isn't allowed at this location. Yet, the Eclipse compiler (ECJ) swallows it just fine. Checkstyle thus fails with the error shown above in an Eclipse environment.

It's clear that the `final` doesn't make any sense at this location. However, if there are compilers out there that accept it, Checkstyle should also accept it.

**Comment by SimplyDanny** [comment]
> Please share javac output also in issue description.

Done.

**Comment by SimplyDanny** [comment]
> Is there a reason you don't just remove the `final` and live without it? Is it providing you something that we aren't seeing?

The problem is that sometimes this is a leftover from code like `n instanceof final Integer i` where the `final` indeed has a semantic meaning. If `i` becomes unused at some point and is removed, people often just forget to remove the `final` as well since the compiler doesn't complain. From that moment on, Checkstyle fails for the whole file and it's hard to figure out why.

**Comment by romani** [comment]
I think it is big in Eclipse compiler, and we should report issue to them.

**Comment by SimplyDanny** [comment]
I have created a [bug](https://github.com/eclipse-jdt/eclipse.jdt.core/issues/1076) in Eclipse. Let's see what the outcome will be there.

In case they happen to let ECJ behave according to the specification (i.e. let it error on this kind of syntax and perhaps even offer an automatic cleanup), a fix can be expected for their 2023-09 release at the earliest. That, however, means Checkstyle opaquely continues to fail for all previous releases containing this syntax. Would that be something you are going to accept?

On the other hand, my fix is very easy in that it lets the parser just swallow an optional `final` token in front of the type. It doesn't even change the generated AST. Thus there isn't an impact on any of Checkstyle's rules whatsoever. So I wonder if it may still be accepted to have a faster solution at hand.

**Comment by nrmancuso** [comment]
Our grammar follows the JLS as closely as possible (while being fully backwards compatible with previous versions of checkstyle). The expression production rule is the most performance sensitive area of our grammar; we should not extend any of it’s alternatives to support a bug in a third party compiler. If this syntax does not exist in the JLS and can’t be compiled by javac, this issue should be closed.

**Comment by romani** [comment]
Do we have a link to java grammar to prove that it is not  supported even at grammar level? 
so it would clearly mean that eclipse compiler (implementation of grammar) just has a gap.

**Comment by nrmancuso** [comment]
Relevant grammar: https://docs.oracle.com/javase/specs/jls/se20/html/jls-15.html#jls-15.20.2

The only mention I find of `final` modifier is in pattern matching, where we also declare a variable (which we do not do in the example in this issue).

**Comment by romani** [comment]
Looks like issue is accepted by eclipse team.

I am closing this issue.
